### PR TITLE
Improve calypso app builds

### PR DESCRIPTION
**Pull request by noahtallen**
### Changes proposed in this Pull Request
Each calypso app has something in common: they are deployed to a location on WordPress.com. Thus far, we have only really supported syncing to WordPress.com for ETK, with support for the other apps often being broken or nonexistent.

The goal of this PR is to make each app behave the same way:
1. `yarn dev` will run webpack, watching for changes.
2. `yarn build` will run webpack in production mode without watching for changes.
3. Adding `--sync` to either of the above commands will sync the changes to the correct location on WordPress.com

#### Unified build script

We accomplish this with a bin script which launches the build commands. Each package gets the following commands:
```json
{
	"build:some-entry": "calypso-build --source='some/entrypoint'",
	"clean": "npx rimraf dist",
	"dev": "node ../../bin/build-calypso-app.mjs --localPath dist --remotePath /home/wpcom/public_html/path",
	"build": "NODE_ENV=production yarn dev",
}
```

An app may have several entrypoints, each specified under a `build:` script. These are all run in parallel by `npm-run-all`. The user can manually specify `--sync`, and this will be passed to the build-calypso-app script.

The builder script first cleans the app, then runs all `build:*` tasks in parallel. It then conditionally syncs to the remote depending on whether `--sync` exists. It syncs once in production mode, and listens for syncs in development mode. The sync command itself is debounced to 500ms to avoid overwhelming the connection while webpack is building.

This approach allows us to simplify the ETK package.json, and it allows us to share the same build script across all calypso apps. This way, a fix to the build script will apply everywhere.

#### Fixes for app webpack builds:
1. Resolves the ETK issue where Node says there is a memory leak because there are too many event emitters. By increasing the event emitter limit to 20, we avoid the problem.
2. Resolves the issue with webpack 5 deprecation warnings. These were coming from the duplicate package plugin, which is already disabled for calypso client builds. I've disabled that by default for all calypso apps. 

#### Output changes:
1. A few ETK files register as "changed". This is because I unified a few `build:` scripts into the same one. `calypso-build` can accept multiple sources, so we didn't need to split these scripts out into separate build tasks. We only need separate build tasks for when the dist folder is separate, which was not the case for the ones I changed.

### Todo:
- [x] Process failure does not close child processes. For example, if rsync fails, the script will exit, but webpack processes will be left dangling. 
- [x] rsync processes may not be getting terminated

### Testing instructions
for each of the wpcom apps:
1. Run `yarn build` and `yarn dev` locally and verify they work properly.
2. Run `yarn dev --sync` and test that syncing to wpcom and watching for changes works properly.
3. Verify that the webpack bundles in TeamCity have not changed in any meaningful way.

### CODE REVIEW

File: bin/build-calypso-app.mjs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		if ( ! timeout ) {
+			func();
+		} else {
+			calledWithTimeout = true;
+		}
+		clearTimeout( timeout );
+		timeout = setTimeout( () => {
+			if ( calledWithTimeout ) {
+				func();
+				calledWithTimeout = false;
```
Comment (jsnajdr) [inline bin/build-calypso-app.mjs]: You need to set `timeout = null` here, too. Otherwise the function will never be called again after the first timeout elapses.

File: apps/notifications/package.json [hunk 2]
```diff
@@ -18,10 +18,12 @@
 	"homepage": "https://github.com/Automattic/wp-calypso",
 	"scripts": {
 		"clean": "npx rimraf dist",
-		"build": "yarn clean && BROWSERSLIST_ENV=wpcom calypso-build --config='./webpack.config.js'",
+		"build:o2-blocks": "calypso-build --config='./webpack.config.js'",
```
Comment (jsnajdr) [inline apps/notifications/package.json]: Should this be `build:notifications`?

File: bin/build-calypso-app.mjs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	// In dev mode, we start watching to sync while the webpack build is happening.
+	if ( watch && sync ) {
+		syncToRemote( localPath, remotePath, true );
+	}
+}
+
+function syncToRemote( localPath, remotePath, watch = false ) {
+	const debouncedSync = debouncer( () => {
+		console.info( 'Performing sync...' );
+		exec( `rsync -ahz --exclude=".*" ${ localPath } wpcom-sandbox:${ remotePath }` );
```
Comment (jsnajdr) [inline bin/build-calypso-app.mjs]: `rsync` uses SSH to connect to the remote host, and uses the host aliases defined in my SSH config, like `wpcom-sandbox`? And the `wpcom-sandbox` name is hardcoded? It should be overridable with an env variable.

About the debouncing logic:
- is it safe to start a new `rsync` even if the previous one hasn't finished yet? Can happen if it takes more than 500ms
- should we really start the `rsync` immediately on first call, without waiting? Then it will start right when webpack emits the very first file, and will sync while webpack is still writing. A better logic would be: after a first write, schedule a sync, and perform it only after writes have ceased for at least 500ms. That's exactly what the plain Lodash `_.debounce( sync, 500 )` would do.

File: packages/calypso-build/webpack.config.js [hunk 4]
```diff
@@ -21,6 +21,7 @@ const { cssNameFromFilename, shouldTranspileDependency } = require( './webpack/u
  */
 const isDevelopment = process.env.NODE_ENV !== 'production';
 const cachePath = path.resolve( '.cache' );
+const shouldCheckForDuplicatePackages = ! process.env.DISABLE_DUPLICATE_PACKAGE_CHECK;
```
Comment (noahtallen) [inline packages/calypso-build/webpack.config.js:24]: Keeps the duplicated package checker enabled by default, but allows us to disable it to avoid webpack 5 deprecation warnings. Inspired by https://github.com/Automattic/wp-calypso/pull/45858

File: bin/build-calypso-app.mjs (hunk 1 above)
Comment (noahtallen) [reply bin/build-calypso-app.mjs]: I think it would still run after the next timeout runs out. The next time the debounced function is called, another timeout will still get set. (it won't get called on the leading edge though!)

that said, we don't need this logic if we stop calling on the leading edge.

I originally added this since I was thinking about https://github.com/Automattic/wp-calypso/issues/48830. But I think those are two different problems

File: bin/build-calypso-app.mjs (hunk 3 above)
Comment (noahtallen) [reply bin/build-calypso-app.mjs]: > is it safe to start a new rsync even if the previous one hasn't finished yet? Can happen if it takes more than 500ms

I've reworked the logic to kill the existing rsync process before trying a new sync. I think this should be ok

> should we really start the rsync immediately on first call, without waiting

Hmm, yeah, we don't have to do an initial sync. What if `chokidar` didn't register any change events after the sync logic got set up? This could theoretically happen if the webpack build finishes before the call to `chokidar`. That does seem pretty unlikely, though.

**Comment by noahtallen** [comment]
Still working on getting the process spawning to work properly. I don't think child process.kill() on rsync is doing quite what I expect.

But a bigger issue is that if rsync fails, then `npm-run-all` will not close its child processes. From what I can tell, when we exit on error, it doesn't clean up its child processes, and apparently neither does Node. As a result, if rsync fails in dev mode, we get a ton of dangling webpack processes. 

Interestingly, it does cleanup correct when exiting with ctrl-c. Just not if we do `process.exit`. I also can't clean them up by doing:

```js
const proc = spawn( 'yarn', [ 'run-p', 'build:*' ] );

process.on('exit', () => {
 proc.kill();
})
```

**Comment by noahtallen** [comment]
> Still working on getting the process spawning to work properly. I don't think child process.kill() on rsync is doing quite what I expect.

Alright, this is resolved now. I've used `tree-kill` on error to make sure there are no dangling processes.

File: apps/notifications/package.json [hunk 5]
```diff
@@ -18,10 +18,11 @@
...
 	"scripts": {
 		"clean": "npx rimraf dist",
-		"build": "yarn clean && BROWSERSLIST_ENV=wpcom calypso-build --config='./webpack.config.js'",
+		"build:notifications": "calypso-build --config='./webpack.config.js'",
 		"dev-server": "webpack serve --inline --progress --liveReload --content-base dist --watch-content-base --host notifications.localhost --port 8888",
-		"start": "yarn run clean && yarn run build && yarn run dev-server",
-		"sync": "yarn run clean && yarn run build && sh ./bin/wpcom-watch-and-sync.sh"
+		"start": "yarn run clean && yarn run build:notifications && yarn run dev-server",
+		"dev": "node ../../bin/build-calypso-app.mjs --localPath dist --remotePath /home/wpcom/public_html/widgets.wp.com/notifications",
+		"build": "NODE_ENV=production yarn dev"
```
Comment (fullofcaffeine) [inline apps/notifications/package.json:25]: Why does `build` call `dev`? Isn't `build` supposed to build for production?

Comment (fullofcaffeine) [reply apps/notifications/package.json:25]: Or maybe I misunderstood the semantics of those commands 🤔

File: apps/notifications/package.json [hunk 6]
```diff
@@ -18,10 +18,11 @@
 	"homepage": "https://github.com/Automattic/wp-calypso",
 	"scripts": {
 		"clean": "npx rimraf dist",
-		"build": "yarn clean && BROWSERSLIST_ENV=wpcom calypso-build --config='./webpack.config.js'",
+		"build:notifications": "calypso-build --config='./webpack.config.js'",
```
Comment (fullofcaffeine) [inline apps/notifications/package.json]: Could you elaborate on those namespaced commands? I see you referenced them as additional entrypoints, but they weren't defined before. Why are those needed and why are they're here now?

Comment (fullofcaffeine) [reply apps/notifications/package.json]: Also, I find it a bit confusing that `build:notifications` calls `calypso-build`, while `build` calls `dev`. Could you elaborate here, too?

**Review by fullofcaffeine — APPROVED** [review]
I followed the test instructions and it worked well! I also left a few non-blocking questions. 

I found the new commands don't apply to the `full-site-editing` app, but I assume that's okay as this app is deprecated (AFAIK)?

File: apps/notifications/package.json (hunk 5 above)
Comment (noahtallen) [reply apps/notifications/package.json:25]: The only difference between the two modes  is `NODE_ENV`. `NODE_ENV` is the only thing which makes a difference either way under the hood. For example, if node env is development (which is the default for these scripts), then webpack will start in watch mode. If it is production, it will run a single build.

**Comment by noahtallen** [comment]
> I found the new commands don't apply to the full-site-editing app, but I assume that's okay as this app is deprecated (AFAIK)?

I think those files were deleted a while ago, maybe they never got cleaned up on your system?

Thanks for the review and the questions!

File: apps/notifications/package.json (hunk 6 above)
Comment (noahtallen) [reply apps/notifications/package.json]: Great question! Here's how it works:

So `calypso-build` is the root build task. It will call webpack and do the actual work to build the plugin. However, calypso build by itself doesn't (and probably shouldn't) include any code which would let us sync anything to a sandbox at the same time. In order to allow the sync operation to happen at the same time, we need to instead call our own script which can orchestrate everything.

Our own script which orchestrates everything will run all build tasks defined by the plugin in parallel, and will also start the sync part automatically when the flag is specified.

So basically:
- Both build and dev call `build-calypso-app.mjs`. (The only difference between build/dev is the NODE_ENV)
- `build-calypso-app.mjs` starts the sync script based on the presence of `--sync`
- `build-calypso-app.mjs` handles runing in watch mode or production mode based on the NODE_ENV.
- `build-calypso-app.mjs` starts all `build:*` tasks defined in package.json in parallel.
- the `build:*` tasks call calypso-build 
- `calypso-build` calls `webpack`

We _could_ have another package script which calls the build-calypso-app script, and then both build and dev call that instead. But IMO it just takes up extra space, since we can just put that call into either build or dev, and then have the other call that.

The script is important because we can't easily orchestrate all this stuff using _just_ package.json scripts (nor can we keep it consistent.) The namespacing for `build:` is important because some apps define multiple webpack builds, all of which need to run together (ideally in parallel). For example, ETK has 11 or 12 different webpack builds to handle each sub-plugin.

Notifications doesn't have multiple webpack builds, so it is a little weird here. But we need some kind of format for the actual webpack build so that our build-calyps-app script knows exactly what to call. (`npm-run-all` can handle wildcards, which lets us use `build:*` to run webpack)

I'm happy for suggestions because I know it's a bit confusing.
